File: Core/services/session_service.py
"""
Session Service — Conversation Management.

Responsibility: Manage conversation history and context.
  - Create sessions
  - Add messages
  - Track conversation state
  - Persist history to disk

Pure Python, type hinted, fully testable.
Adapted from RAG_RAT/Core/services/session_service.py.
"""

from __future__ import annotations

import json
import logging
import uuid
from datetime import datetime
from pathlib import Path
from typing import Any, Dict, List, Optional

from Core.exceptions import ValidationError

logger = logging.getLogger(__name__)


class SessionService:
    """
    Service for session management.

    Pure business logic — no UI dependencies.
    """

    def __init__(self, storage_dir: Optional[Path] = None):
        if storage_dir is None:
            try:
                from config_system import config

                storage_dir = config.BASE_DIR / "conversation_cache"
            except ImportError:
                storage_dir = Path.cwd() / "conversation_cache"
        self.storage_dir = Path(storage_dir)
        self.storage_dir.mkdir(parents=True, exist_ok=True)
        self._active_sessions: Dict[str, Dict[str, Any]] = {}
        logger.info(f"✓ SessionService initialized: {self.storage_dir}")

    def create_session(self, user_id: Optional[str] = None) -> Dict[str, Any]:
        """Create a new session."""
        session_id = str(uuid.uuid4())
        session: Dict[str, Any] = {
            "id": session_id,
            "user_id": user_id or "anonymous",
            "created_at": datetime.now().isoformat(),
            "updated_at": datetime.now().isoformat(),
            "messages": [],
            "metadata": {},
        }
        self._active_sessions[session_id] = session
        self._save_session(session)
        logger.info(f"✓ Created session: {session_id}")
        return session

    def add_message(
        self,
        session_id: str,
        role: str,
        content: str,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """Add a message to a session."""
        if not session_id:
            raise ValidationError("Session ID is required", field="session_id")
        if role not in ("user", "assistant", "system"):
            raise ValidationError(f"Invalid role: {role}", field="role")
        if not content or not content.strip():
            raise ValidationError("Content cannot be empty", field="content")

        session = self._get_or_load(session_id)
        msg: Dict[str, Any] = {
            "role": role,
            "content": content,
            "timestamp": datetime.now().isoformat(),
            "metadata": metadata or {},
        }
        session["messages"].append(msg)
        session["updated_at"] = datetime.now().isoformat()
        self._save_session(session)
        return msg

    def get_history(self, session_id: str, limit: int = 20) -> List[Dict[str, Any]]:
        """Get the last *limit* messages for a session."""
        session = self._get_or_load(session_id)
        return session["messages"][-limit:]

    def clear_session(self, session_id: str) -> bool:
        """Clear all messages in a session."""
        session = self._get_or_load(session_id)
        session["messages"] = []
        session["updated_at"] = datetime.now().isoformat()
        self._save_session(session)
        self._active_sessions.pop(session_id, None)
        logger.info(f"✓ Cleared session: {session_id}")
        return True

    def list_sessions(self) -> List[Dict[str, Any]]:
        """List all sessions (summaries only)."""
        summaries: List[Dict[str, Any]] = []
        for path in self.storage_dir.glob("*.json"):
            try:
                data = json.loads(path.read_text(encoding="utf-8"))
                summaries.append(
                    {
                        "id": data.get("id"),
                        "user_id": data.get("user_id"),
                        "created_at": data.get("created_at"),
                        "updated_at": data.get("updated_at"),
                        "message_count": len(data.get("messages", [])),
                    }
                )
            except Exception:
                continue
        return summaries

    # ─── Private helpers ─────────────────────────────────

    def _get_or_load(self, session_id: str) -> Dict[str, Any]:
        if session_id in self._active_sessions:
            return self._active_sessions[session_id]
        path = self.storage_dir / f"{session_id}.json"
        if path.exists():
            try:
                session = json.loads(path.read_text(encoding="utf-8"))
            except json.JSONDecodeError as exc:
                raise ValidationError(f"Corrupt session file: {session_id}", field="session_id") from exc
            self._active_sessions[session_id] = session
            return session
        raise ValidationError(f"Session not found: {session_id}", field="session_id")

    def _save_session(self, session: Dict[str, Any]) -> None:
        path = self.storage_dir / f"{session['id']}.json"
        path.write_text(json.dumps(session, indent=2), encoding="utf-8")

File: rust_output/src/session_service.rs
/// Session Service — Conversation Management.
/// 
/// Responsibility: Manage conversation history and context.
/// - Create sessions
/// - Add messages
/// - Track conversation state
/// - Persist history to disk
/// 
/// Pure Python, type hinted, fully testable.
/// Adapted from RAG_RAT/Core/services/session_service::py.

use anyhow::{Result, Context};
use crate::exceptions::{ValidationError};
use std::collections::HashMap;
use std::path::PathBuf;

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

/// Service for session management.
/// 
/// Pure business logic — no UI dependencies.
#[derive(Debug, Clone)]
pub struct SessionService {
    pub storage_dir: PathBuf,
    pub _active_sessions: HashMap<String, HashMap<String, Box<dyn std::any::Any>>>,
}

impl SessionService {
    pub fn new(storage_dir: Option<PathBuf>) -> Self {
        Self {
            storage_dir: PathBuf::from(storage_dir),
            _active_sessions: HashMap::new(),
        }
    }
    /// Create a new session.
    pub fn create_session(&mut self, user_id: Option<String>) -> HashMap<String, Box<dyn std::any::Any>> {
        // Create a new session.
        let mut session_id = /* uuid */ "00000000-0000-0000-0000-000000000000".to_string().to_string();
        let mut session = HashMap::from([("id".to_string(), session_id), ("user_id".to_string(), (user_id || "anonymous".to_string())), ("created_at".to_string(), datetime::now().isoformat()), ("updated_at".to_string(), datetime::now().isoformat()), ("messages".to_string(), vec![]), ("metadata".to_string(), HashMap::new())]);
        self._active_sessions[session_id] = session;
        self._save_session(session);
        logger.info(format!("✓ Created session: {}", session_id));
        session
    }
    /// Add a message to a session.
    pub fn add_message(&mut self, session_id: String, role: String, content: String, metadata: Option<HashMap<String, Box<dyn std::any::Any>>>) -> Result<HashMap<String, Box<dyn std::any::Any>>> {
        // Add a message to a session.
        if !session_id {
            return Err(anyhow::anyhow!("ValidationError('Session ID is required', field='session_id')"));
        }
        if !("user".to_string(), "assistant".to_string(), "system".to_string()).contains(&role) {
            return Err(anyhow::anyhow!("ValidationError(f'Invalid role: {role}', field='role')"));
        }
        if (!content || !content.trim().to_string()) {
            return Err(anyhow::anyhow!("ValidationError('Content cannot be empty', field='content')"));
        }
        let mut session = self._get_or_load(session_id);
        let mut msg = HashMap::from([("role".to_string(), role), ("content".to_string(), content), ("timestamp".to_string(), datetime::now().isoformat()), ("metadata".to_string(), (metadata || HashMap::new()))]);
        session["messages".to_string()].push(msg);
        session["updated_at".to_string()] = datetime::now().isoformat();
        self._save_session(session);
        Ok(msg)
    }
    /// Get the last *limit* messages for a session.
    pub fn get_history(&mut self, session_id: String, limit: i64) -> Vec<HashMap<String, Box<dyn std::any::Any>>> {
        // Get the last *limit* messages for a session.
        let mut session = self._get_or_load(session_id);
        session["messages".to_string()][-limit..]
    }
    /// Clear all messages in a session.
    pub fn clear_session(&mut self, session_id: String) -> bool {
        // Clear all messages in a session.
        let mut session = self._get_or_load(session_id);
        session["messages".to_string()] = vec![];
        session["updated_at".to_string()] = datetime::now().isoformat();
        self._save_session(session);
        self._active_sessions.remove(&session_id).unwrap_or(None);
        logger.info(format!("✓ Cleared session: {}", session_id));
        true
    }
    /// List all sessions (summaries only).
    pub fn list_sessions(&mut self) -> Result<Vec<HashMap<String, Box<dyn std::any::Any>>>> {
        // List all sessions (summaries only).
        let mut summaries = vec![];
        for path in self.storage_dir.glob("*.json".to_string()).iter() {
            // try:
            {
                let mut data = serde_json::from_str(&path.read_to_string())).unwrap();
                summaries.push(HashMap::from([("id".to_string(), data.get(&"id".to_string()).cloned()), ("user_id".to_string(), data.get(&"user_id".to_string()).cloned()), ("created_at".to_string(), data.get(&"created_at".to_string()).cloned()), ("updated_at".to_string(), data.get(&"updated_at".to_string()).cloned()), ("message_count".to_string(), data.get(&"messages".to_string()).cloned().unwrap_or(vec![]).len())]));
            }
            // except Exception as _e:
        }
        Ok(summaries)
    }
    pub fn _get_or_load(&mut self, session_id: String) -> Result<HashMap<String, Box<dyn std::any::Any>>> {
        if self._active_sessions.contains(&session_id) {
            self._active_sessions[&session_id]
        }
        let mut path = (self.storage_dir / format!("{}.json", session_id));
        if path.exists() {
            // try:
            {
                let mut session = serde_json::from_str(&path.read_to_string())).unwrap();
            }
            // except json::JSONDecodeError as _e:
            self._active_sessions[session_id] = session;
            session
        }
        return Err(anyhow::anyhow!("ValidationError(f'Session not found: {session_i}', field='session_id')"));
    }
    pub fn _save_session(&mut self, session: HashMap<String, Box<dyn std::any::Any>>) -> () {
        let mut path = (self.storage_dir / format!("{}.json", session["id".to_string()]));
        pathstd::fs::write(&serde_json::to_string(&session).unwrap(), /* encoding= */ "utf-8".to_string());
    }
}
